examples/merkle_tree.rs

```
@@ -0,0 +1,125 @@
+
+	let mut rng = StdRng::seed_from_u64(0);
+	// Create a Merkle tree with 8 leaves
+	let leaves = (0..1 << args.log_leaves)
+		.map(|_| rng.r#gen::<[u8; 32]>())
+		.collect::<Vec<_>>();
+
+	let tree = MerkleTree::new(&leaves);
+
+	let roots = tree.root();
+	let paths = (0..1 << args.log_paths)
+		.map(|_| {
+			let index = rng.gen_range(0..1 << args.log_leaves);
+			MerklePath {
+				root_id: 0,
+				index,
+				leaf: leaves[index],
+				nodes: tree.merkle_path(index),
+			}
+		})
+		.collect::<Vec<_>>();
+
+	let trace_gen_scope = tracing::info_span!("generating trace").entered();
+	let trace = MerkleTreeTrace::generate(vec![roots], &paths);
+
+	// Allocate memory for the witness
+	let allocator = Bump::new();
+	let mut witness = WitnessIndex::<PackedType<OptimalUnderlier, B128>>::new(&cs, &allocator);
+
+	// Fill the tables with the trace
+	merkle_tree_cs.fill_tables(&trace, &cs, &mut witness)?;
+
+	// Create boundary values based on the trace's boundaries
+	let boundaries = merkle_tree_cs.make_boundaries(&trace);
+	drop(trace_gen_scope);
+
+	let table_sizes = witness.table_sizes();
+
+	let statement = Statement::<B128> {
+		boundaries,
+		table_sizes,
+	};
+
+	let ccs = cs.compile(&statement).unwrap();
+	let witness = witness.into_multilinear_extension_index();
+
+	let proof = binius_core::constraint_system::prove::<
+		OptimalUnderlier,
+		CanonicalTowerFamily,
+		Groestl256,
+		Groestl256ByteCompression,
+		HasherChallenger<Groestl256>,
+		_,
+	>(
+		&ccs,
+		args.log_inv_rate as usize,
+		SECURITY_BITS,
+		&statement.boundaries,
+		witness,
+		&binius_hal::make_portable_backend(),
+	)
+	.unwrap();
+
+	println!("Proof size: {}", ByteSize::b(proof.get_proof_size() as u64));
+
+	binius_core::constraint_system::verify::<
+		OptimalUnderlier,
+		CanonicalTowerFamily,
+		Groestl256,
+		Groestl256ByteCompression,
+		HasherChallenger<Groestl256>,
+	>(&ccs, args.log_inv_rate as usize, SECURITY_BITS, &statement.boundaries, proof)
+	.unwrap();
```

nit: you can just reraise the error